resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -0,0 +1,191 @@
+        if (minTtl > maxTtl) {
+            throw new IllegalArgumentException(
+                    "minTtl: " + minTtl + ", maxTtl: " + maxTtl + " (expected: 0 <= minTtl <= maxTtl)");
+        }
+        this.negativeTtl = checkPositiveOrZero(negativeTtl, "negativeTtl");
+    }
+
+    @Override
+    public int minTtl() {
+        return minTtl;
+    }
+
+    @Override
+    public int maxTtl() {
+        return maxTtl;
+    }
+
+    @Override
+    public int negativeTtl() {
+        return negativeTtl;
+    }
+
+    @Override
+    public void clearCache() {
+        for (Iterator<Map.Entry<String, List<DnsCacheEntry>>> i = resolveCache.entrySet().iterator(); i.hasNext();) {
+            final Map.Entry<String, List<DnsCacheEntry>> e = i.next();
+            i.remove();
+            cancelExpiration(e.getValue());
+        }
+    }
+
+    @Override
+    public boolean clearCache(String hostname) {
+        boolean removed = false;
+        for (Iterator<Map.Entry<String, List<DnsCacheEntry>>> i = resolveCache.entrySet().iterator(); i.hasNext();) {
+            final Map.Entry<String, List<DnsCacheEntry>> e = i.next();
+            if (e.getKey().equals(hostname)) {
+                i.remove();
+                cancelExpiration(e.getValue());
+                removed = true;
+            }
+        }
+        return removed;
+    }
+
+    @Override
+    public List<DnsCacheEntry> get(String hostname) {
+        return resolveCache.get(hostname);
+    }
+
+    private List<DnsCacheEntry> cachedEntries(String hostname) {
+        List<DnsCacheEntry> oldEntries = resolveCache.get(hostname);
+        final List<DnsCacheEntry> entries;
+        if (oldEntries == null) {
+            List<DnsCacheEntry> newEntries = new ArrayList<DnsCacheEntry>(8);
+            oldEntries = resolveCache.putIfAbsent(hostname, newEntries);
+            entries = oldEntries != null? oldEntries : newEntries;
+        } else {
+            entries = oldEntries;
+        }
+        return entries;
+    }
+
+    @Override
+    public void cache(String hostname, InetAddress address, long originalTtl, EventLoop loop) {
+        if (maxTtl == 0) {
+            return;
+        }
+
+        final int ttl = Math.max(minTtl, (int) Math.min(maxTtl, originalTtl));
+        final List<DnsCacheEntry> entries = cachedEntries(hostname);
+        final DnsCacheEntry e = new DnsCacheEntry(hostname, address);
+
+        synchronized (entries) {
+            if (!entries.isEmpty()) {
+                final DnsCacheEntry firstEntry = entries.get(0);
+                if (firstEntry.cause() != null) {
+                    assert entries.size() == 1;
+                    firstEntry.cancelExpiration();
+                    entries.clear();
+                }
+            }
+            entries.add(e);
+        }
+
+        scheduleCacheExpiration(entries, e, ttl, loop);
+    }
+
+    @Override
+    public void cache(String hostname, Throwable cause, EventLoop loop) {
```

Comment:
Could you check if `hostname`, `cause` and `loop` are null using `checkNotNull(..., "...")`?